onnx_trt_backend.cpp

```diff
@@ -136,9 +136,34 @@ class OnnxTensorRTEvent {
   cudaEvent_t event_;
 };
 
+class CudaDeviceGuard {
+  public:
+    CudaDeviceGuard(int backend_id) {
+      if (cudaGetDevice(&current_device_) != cudaSuccess) {
+        throw std::runtime_error("Cannot run cudaGetDevice");
+      }
+      if (current_device_ != backend_id) {
+        if (cudaSetDevice(backend_id) != cudaSuccess) {
+          throw std::runtime_error("Cannot run cudaSetDevice");
+        }
+        need_restore_ = true;
+      }
+    }
+
+    ~CudaDeviceGuard() {
+      if (need_restore_) {
+        cudaSetDevice(current_device_);
+      }
+    }
+
+  private:
+    int current_device_{0};
+    bool need_restore_{false};
```

Comment:
You could use device id = -1 to indicate invalid device id, but this is fine with me too